### DISCUSSION TITLE
How to Get Smooth Closed Splines?

### DISCUSSION THREAD

**Discussion by xvlvx0** [mozman/ezdxf#1013] [Q&A] [answered]
Hi All,

Does anyone know if it is possible for a closed spline (the x,y point-sets first and last point are the same) to be fully tangent all around?

Explanation on my needs; in the project I'm working on I generated a b-spline (scipy interpolate), which is outputted as a set of points. For completeness I make sure the first and last point of the array are the same, thus making it a closed loop b-spline. If I then use ezdxf's add_spline() method and provide it these points it creates the spline but the ends are wonky.

Attached two pictures (zoomed in) that show the start- and endpoints aren't tangent, the point with the arrow is the start/endpoint. I hope it is visible enough as it isn't the best example can reproduce now:
![dxf-spline-end-not-tangent-2](https://github.com/mozman/ezdxf/assets/33175714/46218c9c-1f9a-474a-bd47-b66bb7a0090a)
![dxf-spline-end-not-tangent](https://github.com/mozman/ezdxf/assets/33175714/1acb9aca-950d-4724-966f-941573cdfa54)

Best regards,

LVX

**Comment by xvlvx0** [comment]
Just to append to my question, I saw in the documentation the section about start and end tangency, but I'm confused to what it exactly does.
[spline docs](https://ezdxf.readthedocs.io/en/stable/tutorials/spline.html)

spline.dxf.start_tangent = (0, 1, 0)
spline.dxf.end_tangent = (1, 0, 0)

BR,
LVX

**Comment by mozman** [comment]
# Spline Tangents

The tangents (direction vectors) of a spline are the directions of the first derivative of the curve:

```python
import ezdxf
from ezdxf.math import Vec3
from ezdxf.math import estimate_tangents
import numpy as np

def main():
    doc = ezdxf.new()
    msp = doc.modelspace()
    fit_points = Vec3.list(
        [
            (0, 0, 0),
            (1000, 600, 0),
            (1500, 1200, 0),
            (500, 1250, 0),
            (0, 0, 0),
        ]
    )
    spline = msp.add_cad_spline_control_frame(fit_points)

    ct = spline.construction_tool()
    # draw the curve tangents as red lines:
    for t in np.linspace(0, ct.max_t, 30):
        point, derivative = ct.derivative(t, 1)
        msp.add_line(point, point + derivative.normalize(200), dxfattribs={"color": 1})

    doc.saveas("spline.dxf")

if __name__ == "__main__":
    main()
```
![spline7](https://github.com/mozman/ezdxf/assets/4230966/28b91e15-f998-4a58-b114-0ebe2129a687)

To get a smooth closed curve the start- and end tangents have to be set manually when the control points are calculated and they have to point in the same direction:

```python
    # snip -x-x-x-
    t0 = Vec3(1, -1, 0)  # the length (magnitude) of the tangent is not relevant!
    spline = msp.add_cad_spline_control_frame(fit_points, tangents=[t0, t0])
    # snip -x-x-x-
```

![spline8](https://github.com/mozman/ezdxf/assets/4230966/62d81f7e-f258-4a91-a9ea-6996b38307ec)

To avoid guess work the function `ezdxf.math.estimate_tangents` can be used to estimate the start- and end tangents of the curve:

```python
    # snip -x-x-x-
    tangents = estimate_tangents(fit_points)
    # linear interpolation of the first and the last tangent:
    t0 = tangents[0].lerp(tangents[-1], 0.5)
    msp.add_cad_spline_control_frame(fit_points, tangents=[t0, t0])
    # snip -x-x-x-
```

![spline9](https://github.com/mozman/ezdxf/assets/4230966/f7e952be-ca77-4bfa-90ca-3f60610f0d1d)

It is also possible to add the `SPLINE` by fit-points and setting the tangents as DXF attributes:

```python
    spline = msp.add_spline(fit_points)
    spline.dxf.flags = spline.PERIODIC | spline.CLOSED
    spline.dxf.start_tangent = t0
    spline.dxf.end_tangent = t0
```

**Comment by xvlvx0** [comment]
Hi mozman

Thank you for the nice insight in the usage of the library.

I tried implementing your example, but sadly I'm greeted with an error: "TypeError: unsupported operand type(s) for -: 'list' and 'list'" at the line:
`
tangents = estimate_tangents(xy_data_spline)
`


code snippet:
```python
spline = msp.add_spline(xy_data_spline, dxfattribs={"layer": layer_name})
# get the tangents
tangents = estimate_tangents(xy_data_spline)
# linear interpolation of the first and the last tangent:
t0 = tangents[0].lerp(tangents[-1], 0.5)
# create the spline
spline = msp.add_spline(xy_data_spline, dxfattribs={"layer": "splines_2"})
spline.dxf.flags = spline.PERIODIC | spline.CLOSED
spline.dxf.start_tangent = t0
spline.dxf.end_tangent = t0
```

I couldn't spend much time debugging today, I will have a look at it later this week

**Reply by mozman** [reply, marked as answer]
the `estimate_tangents()` function expects `list[Vec3]`

```python
from ezdxf.math import Vec3

# snip -x-x-x-

tangents = estimate_tangents(Vec3.list(xy_data_spline))
```

**Reply by xvlvx0** [reply]
Nice, that did the trick.
